File: solutions/2062/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Mon Dec 20 15:11:42 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <string> 
#include <unordered_map> 
#include <algorithm> 
#include <numeric> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    int countVowelSubstrings(string word) {
        int result = 0; 
        unordered_map<char, int> umap; 
        umap['a'] = 0;
        umap['e'] = 0;
        umap['i'] = 0; 
        umap['o'] = 0;
        umap['u'] = 0;

        auto checkMap = [&]() {
            for (auto& kv : umap) {
                if (kv.second == 0) return false; 
            }
            return true; 
        };
        auto clearMap = [&]() {
            for (auto& kv : umap) kv.second = 0; 
        };

        for (int i = 0; i < word.size(); ++i) {
            if (umap.find(word[i]) == umap.end()) continue; 
            umap[word[i]]++; 
            for (int j = i+1; j < word.size(); ++j) {
                if (umap.find(word[j]) == umap.end()) break;
                umap[word[j]]++;
                if (checkMap()) result++; 
            }
            clearMap(); 
        }

        return result; 
    }
};

TEST(Test2062, SimpleTest) {
    string word = "cuaieuouac"; 
    Solution s; 
    EXPECT_EQ(s.countVowelSubstrings(word), 7);
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: solutions/2062/main.py
from typing import List 

class Solution:
    @staticmethod
    def checkCount(count):
        for c in count:
            if c == 0:
                return False
        return True
    def countVowelSubstrings(self, word: str) -> int:
        result = 0
        count = [0] * 5
        valid = ['a', 'e', 'i', 'o', 'u']
        for i in range(len(word)):
            if word[i] not in valid:
                continue
            count[valid.index(word[i])] += 1
            j = i + 1
            while j < len(word) and word[j] in valid:
                count[valid.index(word[j])] += 1
                j += 1
                if self.checkCount(count):
                    result += 1
            # reset
            count = [0] * 5
        return result

if __name__ == "__main__": 
    word = "cuaieuouac"
    s = Solution()
    assert s.countVowelSubstrings(word) == 7
